decode/proftomd.1
." $Id: proftomd.1,v 1.1 1004/09/16 21:29:04 yoksas Exp $
.TH PROFTOMD 1 "2004/09/16 21:29:04 $"
.SH NAME
proftomd \- Unidata LDM decoder for FSL netCDF wind profiler data
.SH SYNOPSIS
.HP
.ft B
proftomd
.nh
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%[-d\ \fIdirectory\fP]
\%[-f\ \fIfile\fP]
\%\fIpcode\fP
\%\fIschema\fP
\%\fIMD_number\fP
.hy
.ft
.SH DESCRIPTION
.LP
This program is an LDM decoder for McIDAS data of type \fBMD\fP:
it reads a FSL netCDF data product from a pipe or a file
and decodes that data into a McIDAS \fBMD\fP file.
.LP
This program can be invoked by entries in an LDM configuration file
similar to the following:
.sp
.ta 4n +4n +4n
.nf
##
# Examples of decoding FSL hourly summary and 6-minute profiler data
# that do not save broadcast netCDF file
#
# Hourly summary
FSL2  ^FSL\.NetCDF\.NOAAnet\.windprofiler\.01hr
	PIPE	-close
	proftomd -vl logs/ldm-mcidas.log
	-d /var/data/mcidasd U4 WPRO 81

# 6-minute
FSL2  ^FSL\.NetCDF\.NOAAnet\.windprofiler\.06min
	PIPE	-close
	proftomd -vl logs/ldm-mcidas.log
	-d /var/data/mcidasd U2 WPR6 91

.LP
# Examples of decoding FSL hourly summary and 6-minute profiler data
# that save broadcast netCDF file
#
# Hourly summary
FSL2  ^FSL\.NetCDF\.NOAAnet\.windprofiler\.01hr
	PIPE	-close
	proftomd -vl logs/ldm-mcidas.log
	-n /var/data/mcidasd/\1.nc
	-d /var/data/mcidasd U4 WPRO 81
# 6-minute
FSL2  ^FSL\.NetCDF\.NOAAnet\.windprofiler\.06min
	PIPE	-close
	proftomd -vl logs/ldm-mcidas.log
	-n /var/data/mcidasd/\1.nc
	-d /var/data/mcidasd U2 WPR6 91
.fi
.RE
.LP
The first set of two actions execute \fBproftomd\fP on a temporary file that
\fBproftomd\fP creates.  The temporary file is deleted after the
decoding into the McIDAS \fBMD\fP file.  The temporary file is needed since
\fBproftomd\fP can not be run from data piped to its standard input
since netCDFs are, by nature, random access files.
.LP
The second set of two actions demonstrates how to save the broadcast netCDFs
and then decode them into the McIDAS \fBMD\fP files.
.LP
Additional examples of LDM pqact.conf entries can be
found in the file ldm-mcidas-pqact.conf in the \fBetc/\fP subdirectory of the 
ldm-mcidas installation tree.
.fi
.RE
.LP
This program can also be invoked as a stand-alone program, outside the
\fBLDM\fP environment.
.SH OPTIONS
.TP
.BI "-d " datadir
Change to directory \fIdatadir\fP before decoding.  As a consequence, the
output file will appear in that directory and the \fBROUTE.SYS\fP and
\fBSYSKEY.TAB\fP files will be looked for and, if found, used, and updated
in that directory.  NOTE that the McIDAS schema file \fBSCHEMA\f must
exist in this directory for \fIproftomd\f to work.
.TP
.BI "-f " filename
Specifies that \fIfilename\fP is to be used as the source netCDF for decoding.
.TP
.BI "-n " filename
Save the broadcast netCDF into \fIfilename\f and then decode it into the
McIDAS \fBMD\fP file.  NOTE: \fI-n\f is ignored if \fI-f\f is specified.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata  LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata  LDM Site  
Manager's  Guide  for more information on message logging.
.SH NOTES
.LP
Unidata recommends that the following McIDAS-X routing table update commands
be use to setup \fBROUTE.SYS\fP entries used by \fBproftomd\fP (for more
information see the section on \fBROUTE\fP in the McIDAS-X Users Guide):

.nf
ROUTE ADD U2 MD     X   X CC=3 SYS=2041 2042 2043 "FSL2 hourly wind profiler
ROUTE ADD U6 MD     X   X CC=3 SYS=2044 2045 2046 "FSL2 6-minute Wind profiler
ROUTE REL U2 U6
.fi

The McIDAS schema file, \fBSCHEMA\fP, must exist in the directory into
which \fBproftomd\fP creates data files and be readable by the user
running \fBproftomd\fP. \fBproftomd\fP reads information about the form
of the output \fBMD\fP file from \fBSCHEMA\fP.

\fBROUTE.SYS\fP and \fBSYSKEY.TAB\fP must be readable and writable by
the user running \fBproftomd\fP.  

\fBROUTE.SYS\fP will be used by \fBproftomd\fP if it exists in the
directory in which \fBproftomd\fP creates its output \fBMD\fP files (see
\fB-d\fP flag above) and if it is both readable and writable by the
user running the \fBproftomd\fP.

\fBSYSKEY.TAB\fP entries specified by the \fBSYS=\fP keyword listed above
will be updated if \fBSYSKEY.TAB\fP exists in the directory in which
\fBproftomd\fP creates its output \fBMD\fP files (see \fB-d\fP flag above) and
is both readable and writable by the user running \fBproftomd\fP.
.RE
.SH "SEE ALSO"
.LP
.BR ~ldm-mcidas/etc/ldm-mcidas-pqact.conf,
.BR ldmd (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/proftomd.c
/*
**   Copyright 1997, University Corporation for Atmospheric Research
**   Do not redistribute without permission.
**
**   proftomd.c
**      Generic netCDF format profiler decoder. See usage information
**   for options.
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include "ulog.h"

#ifdef NO_ATEXIT
#include <atexit.h>
#endif

#include "mkdirs_open.h"
#include "profiler.h"

extern void  write_mcidas( char *, char *, char *, int, prof_struct * );
extern void *decode_ncprof( char *, long, int * );

#define HEX80	0x80808080
#define MISS	HEX80

static void usage(argv0)
char *argv0;
{
#define USAGE_FMT "\
Usage: %s [flags] pcode mdbase\n\
Flags:\n\
    -h           Print the help file, then exit the program\n\
    -f filename  Input NetCDF file (default is stdin)\n\
    -n filename  Output NetCDF file name (default not saved)\n\
                 [-n is ignored if -f is used]\n\
    -d path      Change to directory <path> before decoding\n\
    -l filename  Log output (default uses local0, '-' to stdout\n\
    -v           Verbose logging\n\
    -x           Debug logging\n\
    -e VAR=value Set an environmental variable in program\n\
Positionals:\n\
    pcode        McIDAS product code (U2 - 6 min, U4 - hourly)\n\
    schema       McIDAS MD file schema (WPR6 - 6 min, WPRO - hourly)\n\
    mdbase       McIDAS base MD file number (91 - 6 min, 81 - hourly)\n"

    (void) fprintf (stderr, USAGE_FMT, argv0);
}

/*
 * Called at exit.
 * This callback routine registered by atexit().
 */
static void cleanup() {
    uinfo("Exiting") ;
    (void) closeulog();
}


typedef struct envlist{
    char *env;
    struct envlist *next;
} envlist;


main(argc,argv)
int argc;
char *argv[];
{

    char          line[255];
    char         *buf;
    char         *datadir = NULL;
    char         *infilnam = NULL;
    char         *mcpathbuf = NULL;       /* MCPATH environment variable ptr */
    char         *outfilnam = NULL;
    char         *logfname = NULL;
    char         *ofil = NULL;
    char         *odir,*tempnam,*dpos;
    char         *pcode, *schema;

    int           in_file;
    int           out_file;
    int           len,mdbase;
    
    envlist *envhead=NULL,*envobj;
    
    extern int    optind;
    extern int    opterr;
    extern char  *optarg;

    int           ch;
    int           logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE));
    
    int           rlen;
    int           DO_REMOVE = 0;
    
    mode_t        omode=S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH;
    
    prof_struct  *head=NULL;
    int           iret;
    
    /*
     * register exit handler
     */
    if ( atexit(cleanup) != 0 ) {
      serror("atexit");
      exit(1);
    }
    
    /*
    ** Parse command line flags
    */
    opterr = 1;
    while ((ch = getopt(argc,argv,"hvxl:d:f:n:e:")) != EOF) {
       switch(ch) {
          case 'h':
             usage(argv[0]);
             exit(0);
             break;
          case 'd':
             datadir = optarg;
             break;
          case 'f':
             infilnam = optarg;
             break;
          case 'n':
             outfilnam = optarg;
             break;
          case 'l':
             logfname = optarg;
             break;
          case 'v':
             logmask |= LOG_MASK(LOG_INFO);
             break;
          case 'x':
             logmask |= LOG_MASK(LOG_DEBUG);
             break;
          case 'e':
             envobj = (envlist *)malloc(sizeof(envlist));
             envobj->env = malloc(strlen(optarg));
             strcpy(envobj->env,optarg);
             envobj->next = envhead;
             envhead = envobj;
             break;
          case '?':
             usage(argv[0]);
             exit(0);
          }
       }
    
    /*
    ** Parse command line positional parameters
    */
    switch(argc - optind) {
       case 3:
          pcode  = argv[optind++];
          schema = argv[optind++];
          mdbase = atol(argv[optind++]);
          break;
       default:
          usage(argv[0]);
          exit(-1);
    }
    
    /*
    ** Update directory for temporary file if 'datadir' specified by user
    */
    if ( datadir != (char *)NULL ) {
      odir = (char *) malloc(strlen(datadir)+20);
      odir[0] = '\0';
      strcat(odir,datadir);
      strcat(odir,"/");
      ofil = odir;
    }
    
    /*
    ** Set ulog mask and announce start
    */
    (void) setulogmask(logmask);
    (void) openulog(ubasename(argv[0]), (LOG_CONS|LOG_PID), LOG_LOCAL0, logfname) ;
    uinfo("Starting up");
    
	/*
	 * Unset MCPATH environment variable if it exists
	 */
	mcpathbuf = getenv("MCPATH");
	if (mcpathbuf != NULL) {
	  uinfo("unsetting MCPATH environment variable");
	  (void) putenv("MCPATH=");
	}

    /*
    ** Set environment variable if '-e' flag set by user
    */
    envobj = envhead;
    while (envobj != NULL) {
      udebug("putenv %s\0",envobj->env);
      if (putenv(envobj->env) != 0)
        uerror("Can't set environmental variable %s\0",envobj->env);
      envobj = envobj->next;
    }
    
    /*
    ** if input from stdin, redirect to netcdf file "outfilnam"
    */
    if (infilnam == 0) {
       if (outfilnam == 0) { /* create a temporary file name */
          tempnam = (char *)malloc(19);
          tempnam[0] = '\0';
          sprintf(tempnam,".tmp_netcdf.XXXXXX\0");
          mktemp(tempnam);
          udebug("creating temporary netcdf file %s\0",tempnam);
          /* create temp file in directory of output  (get path to ofil)*/
          dpos = strrchr(ofil,'/');
          if (dpos == NULL)
             outfilnam = tempnam;
          else {
             outfilnam = (char *)malloc(strlen(ofil)+20);
             outfilnam[0] = '\0';
             strncat(outfilnam,ofil,dpos-ofil);
             strncat(outfilnam,"/",1);
             strcat(outfilnam,tempnam);
          }
          DO_REMOVE = 1;
          udebug("look %s\0",outfilnam);
       }
    
       in_file = STDIN_FILENO;
       out_file = mkdirs_open(outfilnam,O_WRONLY|O_CREAT|O_TRUNC,omode);
       if(out_file < 0) {
          uerror("could not open %s\n",outfilnam);
          exit(-1);
       }

       buf = (char *)malloc(8196);
       while((rlen = read(in_file,buf,8196)) > 0) {
          udebug("writing %d bytes\0",rlen);
          write(out_file,buf,rlen);
       }
       free(buf);
       close(out_file);
       infilnam = outfilnam;
       udebug("now will input netcdf from %s\0",infilnam);
    } else {
       udebug("will get input from %s\0",infilnam);
    }
    
    /*
    ** call netcdf decoder with infilnam
    */
    head = (prof_struct *) decode_ncprof(infilnam,MISS,&iret);
    
    if ( (head != NULL) && (iret == 0) )
       write_mcidas(datadir, pcode, schema, mdbase, head);
    else
       uerror("could not decode data %d\0",iret);
    
    /*
    ** <<<<< UPC move 20021221 >>>>>
    ** we are done with netcdf file now, remove if necessary
    */
    if ( DO_REMOVE == 1 ) { 
      udebug( "deleting temporary netcdf file %s\0", outfilnam );
      if ( unlink(outfilnam) != 0 )
        uerror( "Could not delete temp file %s\0", outfilnam );
    }
    
    /*
    ** Done
    */
    exit(0);
}
